### DISCUSSION TITLE
Manually building stream formats

### DISCUSSION THREAD

**Discussion by degol** [tomaka/cpal#1315] [Q&A] [answered]
Hello everyone, 

I’m new to CPAL, and I’m trying to port an audio processing application to rust based web application. I have a couple of questions that I couldn’t find the answers of which in the documentation provided.

1. In the machine that I’m using, the possible buffer-size range is set to “unknown”, but when I set it to any value, including the default values, I keep getting the following error: 
  `Error: The requested stream configuration is not supported by the device.`
  
  This is how I'm manually configuring it
  ```
      let sample_rate: cpal::SampleRate = SampleRate{0: 44100 as u32};
      let buffer_size: cpal::BufferSize = BufferSize::Default;
      let config: cpal::StreamConfig =  StreamConfig { channels : 1, sample_rate, buffer_size };
  ``` 

2. Is there any way to get pure raw data other than PCM ? Right now I need the pure sampled raw data, before its converted to PCM

It would be really great if anyone can drop a helpful link here.

Thank you!

.

**Comment by pietervandermeer** [comment]
PCM, or Pulse Code Modulation, is just the standard encoding of all raw sample data. I.e. I think you already got what you want? Otherwise you can record it to analog tape maybe if you don't like PCM ;)

Jokes aside, you could try other sample rates and try 2 channels. I think most devices are stereo, and mono would incurs additional processing. Some underlying API's may not support that.

**Comment by roderickvd** [comment, marked as answer]
1. Nowadays both `webaudio` and `audioworklet` report supported buffer size ranges that you can configure with `BufferSize::Fixed`.

2. I don't see why you'd want raw `u8` instead of `Sample` but nothing is stopping you from calling `to_ne_bytes`.
